File: go-grpc/replicant/replicant.go
package replicant

import (
	"github.com/psu-csl/replicated-store/go/config"
	"github.com/psu-csl/replicated-store/go/kvstore"
	consensusLog "github.com/psu-csl/replicated-store/go/log"
	"github.com/psu-csl/replicated-store/go/multipaxos"
	pb "github.com/psu-csl/replicated-store/go/multipaxos/comm"
	logger "github.com/sirupsen/logrus"
	"net"
	"strconv"
	"strings"
	"sync"
)

type Replicant struct {
	id            int64
	log           *consensusLog.Log
	multipaxos    *multipaxos.Multipaxos
	ipPort        string
	acceptor      net.Listener
	clientManager *ClientManager
	wg            *sync.WaitGroup
}

func NewReplicant(config config.Config, join bool) *Replicant {
	r := &Replicant{
		id:     config.Id,
		ipPort: config.Peers[config.Id],
		wg:     &sync.WaitGroup{},
	}
	r.log = consensusLog.NewLog(kvstore.CreateStore(config))
	r.multipaxos = multipaxos.NewMultipaxos(r.log, config, join)
	r.clientManager = NewClientManager(r.id, int64(len(config.Peers)), r.multipaxos)
	return r
}

func (r *Replicant) Start() {
	r.multipaxos.Start()
	r.StartExecutorThread()
	r.StartServer()
	r.wg.Wait()
}

func (r *Replicant) Stop() {
	r.StopServer()
	r.multipaxos.Stop()
	r.StopExecutorThread()
}

func (r *Replicant) StartServer() {
	pos := strings.Index(r.ipPort, ":")
	if pos == -1 {
		panic("no separator : in the acceptor port")
	}
	pos += 1
	port, err := strconv.Atoi(r.ipPort[pos:])
	if err != nil {
		panic("parsing acceptor port failed")
	}
	port += 1

	acceptor, err := net.Listen("tcp", ":"+strconv.Itoa(port))
	if err != nil {
		logger.Fatalln(err)
	}
	logger.Infof("%v starting server at port %v\n", r.id, port)
	r.acceptor = acceptor
	r.AcceptClient()
}

func (r *Replicant) StopServer() {
	r.acceptor.Close()
	r.clientManager.StopAll()
}

func (r *Replicant) StartExecutorThread() {
	logger.Infof("%v starting executor thread\n", r.id)
	r.wg.Add(1)
	go r.executorThread()
}

func (r *Replicant) StopExecutorThread() {
	logger.Infof("%v stopping executor thread\n", r.id)
	r.log.Stop()
}

func (r *Replicant) StartRpcServer() {
	r.multipaxos.StartRPCServer()
}

func (r *Replicant) executorThread() {
	for {
		instance := r.log.ReadInstance()
		if instance == nil {
			break
		}
		if instance.Command.Type == pb.CommandType_ADDNODE || instance.
			Command.Type == pb.CommandType_DELNODE {
			r.multipaxos.Reconfigure(instance.Command)
			client := r.clientManager.Get(instance.ClientId)
			if client != nil {
				client.Write("joined")
			}
		} else {
			id, result := r.log.Execute(instance)
			client := r.clientManager.Get(id)
			if client != nil {
				client.Write(result.Value)
			} else if instance.ClientId == -1 {
				values := strings.Split(instance.Command.Value, " ")
				posVarSum, _ := strconv.ParseFloat(values[0], 64)
				negVarSum, _ := strconv.ParseFloat(values[1], 64)
				r.multipaxos.TriggerElection(instance.GetBallot(), posVarSum,
					negVarSum)
			}
		}
	}
	r.wg.Done()
}

func (r *Replicant) AcceptClient() {
	for {
		conn, err := r.acceptor.Accept()
		if err != nil {
			logger.Info(err)
			break
		}
		r.clientManager.Start(conn)
	}
}

func (r *Replicant) Monitor() {
	r.multipaxos.Monitor()
}

func (r *Replicant) TriggerElection() {
	r.multipaxos.TriggerElection(1000, 0, 0)
}

File: c++-tcp/replicant.cc
#include <glog/logging.h>
#include <asio.hpp>
#include <condition_variable>
#include <cstring>
#include <iostream>
#include <memory>
#include <optional>

#include "json.h"
#include "replicant.h"

using asio::ip::tcp;
using nlohmann::json;

Replicant::Replicant(asio::io_context* io_context, json const& config)
    : id_(config["id"]),
      log_(kvstore::CreateStore(config)),
      multi_paxos_(&log_, config, io_context),
      ip_port_(config["peers"][id_]),
      io_context_(io_context),
      acceptor_(asio::make_strand(*io_context_)),
      client_manager_(id_, config["peers"].size(), &multi_paxos_,
                      true, config["threadpool_size"]),
      peer_server_(id_, ip_port_, io_context_,
                   config["peers"].size(), config["threadpool_size"]) {}

void Replicant::Start() {
  peer_server_.StartServer(&multi_paxos_);
  multi_paxos_.Start();
  StartExecutorThread();
  StartServer();
}

void Replicant::Stop() {
  peer_server_.StopServer();
  StopServer();
  StopExecutorThread();
  multi_paxos_.Stop();
}

void Replicant::StartServer() {
  auto pos = ip_port_.find(":") + 1;
  CHECK_NE(pos, std::string::npos);
  int port = std::stoi(ip_port_.substr(pos)) + 1;

  server::CreateListener(port, acceptor_);
  DLOG(INFO) << id_ << " starting server at port " << port;

  auto self(shared_from_this());
  asio::dispatch(acceptor_.get_executor(), [this, self] { 
    server::AcceptClient(io_context_, &client_manager_, &acceptor_); 
  });
}

void Replicant::StopServer() {
  auto self(shared_from_this());
  asio::post(acceptor_.get_executor(), [this, self] { acceptor_.close(); });
  client_manager_.StopAll();
}

void Replicant::StartExecutorThread() {
  DLOG(INFO) << id_ << " starting executor thread";
  executor_thread_ = std::thread(&Replicant::ExecutorThread, this);
}

void Replicant::StopExecutorThread() {
  DLOG(INFO) << id_ << " stopping executor thread";
  log_.Stop();
  executor_thread_.join();
}

void Replicant::ExecutorThread() {
  for (;;) {
    auto r = log_.Execute();
    if (!r)
      break;
    auto [id, result] = std::move(*r);
    auto client = client_manager_.Get(id);
    if (client)
      client->Write(result.value_);
  }
}
